build.sh

```diff
@@ -10,41 +15,49 @@ exit_on_error() {
+  FORCE=1
+fi
+
 npm ci --legacy-peer-deps
-rm -Rf dist build
-node ./compile-ddocs.js
-
-mkdir -p ext/xsl
-mkdir -p ext/enketo-transformer/xsl
-cp ./node_modules/cht-core-4-0/api/src/xsl/openrosa2html5form.xsl ext/xsl
-cp ./node_modules/cht-core-4-0/api/src/enketo-transformer/xsl/* ext/enketo-transformer/xsl
-
-dirs=($(find node_modules/cht-* -maxdepth 0 -type d))
-for dir in "${dirs[@]}"; do
-  (cd "$dir"/webapp && npm ci --legacy-peer-deps --production)
-  (cd "$dir"/api && npm ci --legacy-peer-deps --production)
-  (cd "$dir"/shared-libs/calendar-interval && npm ci --legacy-peer-deps)
-  (cd "$dir"/shared-libs/rules-engine && npm ci --legacy-peer-deps)
-  (cd "$dir"/shared-libs/phone-number && npm ci --legacy-peer-deps --production)
-
-  # patch the daterangepicker for responsiveness
-  # https://github.com/dangrossman/bootstrap-daterangepicker/pull/437
-  (cd "$dir" && patch -f webapp/node_modules/bootstrap-daterangepicker/daterangepicker.js < webapp/patches/bootstrap-daterangepicker.patch)
+rm -Rf build
+
+if [[ 1 == "$FORCE" ]]; then
+  rm -Rf dist
+else
+  for item in `ls dist | grep -v cht-core`; do
+      rm -rf dist/"$item"
+  done
+fi
+
+for version in "${!cht_versions[@]}"; do
+  if [[ ! 1 == "$FORCE" ]] && [ -d dist/"$version" ]; then
+    printf "\033[0;32m== SKIPPING $version ==\n"
+    continue
+  fi
 
-  # 210 patch to disable db-object-widget
-  (cd "$dir" && patch -f webapp/src/js/enketo/widgets.js < ../../patches/210-disable-db-object-widgets.patch)
+  git clone https://github.com/medic/cht-core.git build/"$version"
+  (cd build/"$version" && git reset --hard "${cht_versions[$version]}")
+  (cd build/"$version" && git clean -df)
 
-  # patch enketo to always mark the /inputs group as relevant
-  (cd "$dir" && patch -f webapp/node_modules/enketo-core/src/js/form.js < webapp/patches/enketo-inputs-always-relevant_form.patch)
-  (cd "$dir" && patch -f webapp/node_modules/enketo-core/src/js/relevant.js < webapp/patches/enketo-inputs-always-relevant_relevant.patch)
+  (cd build/"$version" && npm ci)
 
-  # patch enketo to fix repeat name collision bug - this should be removed when upgrading to a new version of enketo-core
-  # https://github.com/enketo/enketo-core/issues/815
-  (cd "$dir" && patch -f webapp/node_modules/enketo-core/src/js/calculate.js < webapp/patches/enketo-repeat-name-collision.patch)
+  node ./compile-ddocs.js "$version"
 
-  # patch messageformat to add a default plural function for languages not yet supported by make-plural #5705
-  (cd "$dir" && patch -f webapp/node_modules/messageformat/lib/plurals.js < webapp/patches/messageformat-default-plurals.patch)
+  (cd build/"$version"/api && npm ci --legacy-peer-deps --production)
+  (cd build/"$version" && patch -f api/src/services/generate-xform.js < ../../patches/generate-xform.patch)
+  # 210 patch to disable db-object-widget
+  (cd build/"$version" && patch -f webapp/src/js/enketo/widgets.js < ../../patches/210-disable-db-object-widgets.patch)
```

Comment:
For the record, we do not need to run the patches from `webapp/patches` anymore since those get applied automatically when building the cht-form component!

These custom patches that are left are what is still needed to make the code work with the test harness.

Reply:
Right!  Really,the stuff being patched up here should probably be handled somehow in the cht-form code anyway.  For now, keeping these remaining patches seemed like the cleanest way to keep everything working in the near-term.